### PR TITLE
Public profiles WIP

### PR DESCRIPTION
**Pull request by Josemanki**
Leaving work to date on public profiles

* Enables a /user/ route followed with a username to pull all sets from that user
* Most translations so far done
* Creates a graphql query for the feature
* Makes a general component for MyBuilds and UserCustomSets
* Change MyBuilds to use buildList  component and query
* Add link to user's profile on builds
* Change public builds' header to show Owner outside of the tooltip
* Change profile picture
* Localization of latest features

### CODE REVIEW

File: client/graphql/queries/userProfileSets.graphql [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+          tagAssociations {
+            id
+            associationDate
+            customSetTag {
+              id
+              name
+              imageUrl
+            }
+          }
+        }
```
Comment (lightyearsaway) [inline client/graphql/queries/userProfileSets.graphql:41]: If this is exactly the same as the "myCustomSets" query, it would be good to put it into a fragment. Or even change that MyBuilds component to just use this query with the current user's username.

File: .prettierrc [hunk 2]
```diff
@@ -1,4 +1,5 @@
 {
+  "printWidth": 80,
```
Comment (lightyearsaway) [inline .prettierrc:2]: Why is this necessary? Looks like this is the default.

File: client/components/common/UserCustomSets.tsx [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+import customSetTagsQuery from 'graphql/queries/customSetTags.graphql';
+import classesQuery from 'graphql/queries/classes.graphql';
+import { classes } from 'graphql/queries/__generated__/classes';
+import { TabPane } from 'rc-tabs';
+
+const PAGE_SIZE = 20;
+const THRESHOLD = 600;
+
+interface Props {
+  username: string | null;
```
Comment (lightyearsaway) [inline client/components/common/UserCustomSets.tsx:39]: Shouldn't username be required? If it's not provided, we shouldn't be rendering this component at all, right?

File: client/components/common/UserCustomSets.tsx [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+          )}`}</h1>
+          <span css={{ color: '#8b949e' }}>Member since 05/01/2021</span>
+        </div>
+      </div>
+      <Tabs
+        defaultActiveKey="1"
+        onChange={() => {}}
+        css={{ gridArea: '4/1/4/3', width: '100%', padding: 10 }}
+      >
+        <TabPane tab="Builds" key="1">
```
Comment (lightyearsaway) [inline client/components/common/UserCustomSets.tsx:210]: "Builds" is a user-facing string that needs i18n, right?

File: client/components/common/UserCustomSets.tsx [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+          <InfiniteScroll
```
Comment (lightyearsaway) [inline client/components/common/UserCustomSets.tsx:211]: Seems like a very large chunk of this page is the same as the MyBuilds component. Can we share more code between them? Maybe have a <Builds /> component that takes in the query result and an onLoadMore as a prop?

File: client/components/common/UserCustomSets.tsx [hunk 6]
```diff
@@ -0,0 +1,466 @@
...
+    },
+    [debouncedSearch, setSearch],
+  );
+
+  const { data: userBuilds, loading: queryLoading, fetchMore } = useQuery<
+    userProfileSets,
+    userProfileSetsVariables
+  >(userProfileSetsQuery, {
+    variables: {
+      username: username as string,
```
Comment (lightyearsaway) [inline client/components/common/UserCustomSets.tsx:73]: You shouldn't need this if you require username.

File: server/oneoff/database_setup.py [hunk 7]
```diff
@@ -33,6 +33,7 @@
 from oneoff.sync_spell import create_spell_stats
 from oneoff.sync_set import create_set
 import oneoff.sync_item
+from oneoff.sync_custom_set_tag import load_and_create_all_custom_set_tags
```
Comment (lightyearsaway) [inline server/oneoff/database_setup.py:36]: I fixed this when I setup my dev environment like a week ago, see line 32

**Review by lightyearsaway — CHANGES_REQUESTED** [review]
Hey, thanks so much for the PR! This is a great start. Added some comments, mostly about product features and sharing more code.

**Comment by lightyearsaway** [comment]
Fixes: #340

File: server/oneoff/database_setup.py (hunk 7 above)
Comment (Josemanki) [reply server/oneoff/database_setup.py:36]: Yup. That slipped into the PR, won't be on the final one 👍

File: .prettierrc (hunk 2 above)
Comment (Josemanki) [reply .prettierrc:2]: Slipped too, but it didn't seem to be my default, files were being saved funny so I manually added it there

File: client/pages/user/[username].tsx [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+  return (
+    <>
+      <Head>
+        <style
+          type="text/css"
+          // eslint-disable-next-line react/no-danger
+        />
+        <title>
+          {getTitle(
+            t('USER_SETS', { username: capitalize(username as string) }),
```
[the remark this answers is not in the document]
Comment (Josemanki) [reply client/pages/user/[username].tsx:32]: I'd argue that for styling purposes we should capitalize, it just looks better I guess? What do you think?

The title is changed now though.

File: client/pages/user/[username].tsx [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    ? router.query.username[0]
+    : router.query.username || null;
+
+  return (
+    <>
+      <Head>
+        <style
+          type="text/css"
+          // eslint-disable-next-line react/no-danger
+        />
```
[the remark this answers is not in the document]
Comment (Josemanki) [reply client/pages/user/[username].tsx:29]: Nothing at all lol, probably ended up there from a test, solved

File: client/components/common/UserCustomSets.tsx [hunk 10]
```diff
@@ -272,7 +272,7 @@ const UserCustomSets: React.FC<Props> = ({ username }) => {
                 ]}
                 showSearch
                 filterOption={(input, option) => {
-                  return (option?.children[1] as string)
+                  return (option?.children[1]! as string)
```
Comment (Josemanki) [inline client/components/common/UserCustomSets.tsx:275]: This slipped - don't mind it

File: client/pages/user/[username].tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (lightyearsaway) [reply client/pages/user/[username].tsx]: Thanks! I'd again support convention with popular websites like Reddit and GitHub, which allow usernames to be lowercase and show it as the user intended everywhere.

File: client/components/common/MyBuilds.tsx [hunk 11]
```diff
@@ -2,498 +2,19 @@
```
Comment (lightyearsaway) [inline client/components/common/MyBuilds.tsx:2]: I feel like you don't need this component anymore - it's totally identical to BuildList right now.

File: client/components/common/UserCustomSets.tsx [hunk 12]
```diff
@@ -196,269 +63,11 @@ const UserCustomSets: React.FC<Props> = ({ username }) => {
             },
           }}
         >
-          <h1 css={{ margin: '10px 0px 0px 0px', fontSize: 28 }}>{`${capitalize(
-            username!,
-          )}`}</h1>
+          <h1 css={{ margin: '10px 0px 0px 0px', fontSize: 28 }}>{username}</h1>
```
Comment (lightyearsaway) [inline client/components/common/UserCustomSets.tsx:66]: I've always tried to stick with multiples of 4 for margins, and `rem` for font sizes. Also, wouldn't `marginTop` here be sufficient?

File: client/components/common/UserCustomSets.tsx [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+            flexDirection: 'column',
+            justifyContent: 'center',
+            marginLeft: 20,
+            [mq[1]]: {
+              marginLeft: 0,
+            },
+          }}
+        >
+          <h1 css={{ margin: '10px 0px 0px 0px', fontSize: 28 }}>{username}</h1>
+          <span css={{ color: '#8b949e' }}>Member since 05/01/2021</span>
```
Comment (lightyearsaway) [inline client/components/common/UserCustomSets.tsx:67]: Can we fill this with real data and translation strings?

File: client/pages/my-builds.tsx [hunk 14]
```diff
@@ -32,7 +32,7 @@ const MyBuildsPage: NextPage = () => {
         />
         <title>{getTitle(t('MY_BUILDS'))}</title>
       </Head>
-      <MyBuilds isMobile />
+      <MyBuilds username={data?.currentUser?.username!} />
```
Comment (lightyearsaway) [inline client/pages/my-builds.tsx:35]: I think you could just do `{data?.currentUser && <MyBuilds username={data.currentUser.username} />}` since this page is meaningless if the user is not logged in and the useEffect should quickly route them somewhere else.

**Review by lightyearsaway — CHANGES_REQUESTED** [review]

File: client/components/common/UserCustomSets.tsx (hunk 13 above)
Comment (lightyearsaway) [reply client/components/common/UserCustomSets.tsx:67]: Also, what color is this? Can we use one of the [antd colors](https://ant.design/docs/spec/colors) in `mixins.ts` (add yourself if necessary)?

File: client/graphql/fragments/userSets.graphql [hunk 15]
```diff
@@ -0,0 +1,40 @@
+fragment userSets on User {
```
Comment (lightyearsaway) [inline client/graphql/fragments/userSets.graphql:1]: I know we use `customSet` and `build` kind of interchangeably in the code, but could we stick to one of those and not add `sets` to the mix, i.e. rename this as `userCustomSet` or `userBuild`? Also, I think you could just have a `abbreviatedCustomSet` fragment or something like that instead of having this only be useful for a user's custom sets (so, everything inside `node` without the pageInfo).

File: client/graphql/queries/userProfileSets.graphql [hunk 16]
```diff
@@ -0,0 +1,14 @@
+#import "graphql/fragments/userSets.graphql"
+
+query userProfileSets(
```
Comment (lightyearsaway) [inline client/graphql/queries/userProfileSets.graphql:3]: Maybe we could just make this `userProfile` which gets all the info required for their profile, including any future additions like folders.

File: client/components/common/BuildList.tsx [hunk 17]
```diff
@@ -0,0 +1,559 @@
...
+                  flex: 1,
+                  ...inputFontSize,
+                  '.ant-input': inputFontSize,
+                  height: 32,
+                }}
+                onChange={onSearch}
+                placeholder={t('SEARCH')}
+              />
+            </div>
+            {classSelect}
```
[the remark this answers is not in the document]
Comment (Josemanki) [reply client/components/common/BuildList.tsx:287]: Screenshot 2022-01-06 at 18 27 40
Screenshot 2022-01-06 at 18 28 45

Looking golden on both views

[the remark this answers is not in the document]
Comment (Josemanki) [reply client/components/common/BuildList.tsx:287]: I just noticed there's a bit of a weird padding on the filters, it's 10px so barely unnoticeable, but asymmetrical. Fixed in the new code.

File: client/components/common/UserCustomSets.tsx (hunk 12 above)
Comment (Josemanki) [reply client/components/common/UserCustomSets.tsx:66]: Changed the margin to 12px, and font size to 1.6rem. `marginTop` alone wouldn't work because the default margins on an h1 element will move the _Member since_ too low, so I explicitly had to make the other margins 0px.

File: client/components/common/ProfilePictureModal.tsx [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+const ProfilePictureModal: React.FC<Props> = ({
+  username,
+  onCancel,
+  visible,
+  currentlyActive,
+}) => {
+  const { t } = useTranslation('common');
+  const [active, setActive] = useState<any>('');
```
Comment (lightyearsaway) [inline client/components/common/ProfilePictureModal.tsx:35]: Why aren't you able to type this?

File: client/components/common/ProfilePictureModal.tsx [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+
+  useEffect(() => {
+    setActive(currentlyActive);
```
Comment (lightyearsaway) [inline client/components/common/ProfilePictureModal.tsx:38]: Why not initialize it as currentlyActive?

File: client/components/common/ProfilePictureModal.tsx [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+            maxHeight: 'auto',
+          },
+        }}
+      >
+        {PROFILE_PICTURES.map((item, index) => {
+          const activeStyle =
+            active === item
+              ? {
+                  borderRadius: 4,
+                  border: `2px solid ${blue6}`,
```
Comment (lightyearsaway) [inline client/components/common/ProfilePictureModal.tsx:130]: I think there are theme-agnostic colors so we can add light theme more easily when we get there.

File: client/graphql/mutations/changeProfilePicture.graphql [hunk 21]
```diff
@@ -0,0 +1,5 @@
+mutation changeProfilePicture($picture: String!) {
+  changeProfilePicture(picture: $picture) {
+    ok
```
Comment (lightyearsaway) [inline client/graphql/mutations/changeProfilePicture.graphql:3]: If you changed this query to return the user with id and profile picture, then you wouldn't need to refetch those other queries like user profile.

File: client/graphql/fragments/userCustomSets.graphql [hunk 22]
```diff
@@ -0,0 +1,42 @@
+fragment userCustomSets on User {
```
Comment (lightyearsaway) [inline client/graphql/fragments/userCustomSets.graphql:1]: I still prefer a fragment that's reusable in more situations, since this is only usable on the user node. Meaning, I think the fragment would be more reusable if it was only what's inside node {}.

File: client/graphql/queries/userProfile.graphql [hunk 23]
```diff
@@ -0,0 +1,6 @@
+query userProfile($username: String!) {
+  userByName(username: $username) {
+    profilePicture
```
Comment (lightyearsaway) [inline client/graphql/queries/userProfile.graphql:3]: Don't forget to add id.

File: client/pages/user/[username].tsx [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+  const { data: currentUser } = useQuery<currentUser>(currentUserQuery);
+
+  const { data: userProfileData } = useQuery<userProfile, userProfileVariables>(
+    userProfileQuery,
+    { variables: { username } },
+  );
+
+  React.useEffect(() => {
+    if (userProfileData?.userByName === null) {
+      router.push('/');
```
Comment (lightyearsaway) [inline client/pages/user/[username].tsx:40]: I think this should return a 404.

**Review by lightyearsaway — CHANGES_REQUESTED** [review]

File: client/components/common/ProfilePictureModal.tsx (hunk 18 above)
Comment (Josemanki) [reply client/components/common/ProfilePictureModal.tsx:35]: Was experimenting with which value I actually wanted to use, ended up with string and forgot to change typing haha.

File: client/components/common/ProfilePictureModal.tsx (hunk 19 above)
Comment (Josemanki) [reply client/components/common/ProfilePictureModal.tsx:38]: Yesterday that approach didn't give me the result I was looking for and today somehow worked lol. Fixed.

File: client/graphql/mutations/changeProfilePicture.graphql (hunk 21 above)
Comment (Josemanki) [reply client/graphql/mutations/changeProfilePicture.graphql:3]: I have gone around one of the refetches by returning the user, refetching the profile is not needed anymore, but buildList is in another component, so I probably should keep that refetch in there? When I return the user I cannot get and repopulate userCustomSets, can I?

File: client/components/common/DeleteCustomSetModal.tsx [hunk 25]
```diff
@@ -30,7 +30,7 @@ const DeleteCustomSetModal: React.FC<Props> = ({
     deleteCustomSetVariables
   >(deleteCustomSetMutation, {
     variables: { customSetId },
-    refetchQueries: () => ['myCustomSets'],
+    refetchQueries: () => ['buildList'],
     awaitRefetchQueries: true,
```
Comment (lightyearsaway) [inline client/components/common/DeleteCustomSetModal.tsx:34]: We can remove this now, right?

File: server/app/migrations/versions/b9b68fd74d2a_.py [hunk 26]
```diff
@@ -0,0 +1,35 @@
...
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.add_column(
+        "user_account",
+        sa.Column("profile_picture", sa.String(length=120), nullable=True),
+    )
+    op.execute(
+        "UPDATE user_account SET profile_picture = 'profile-pictures/ProPic_Emerald_1.png'"
```
Comment (lightyearsaway) [inline server/app/migrations/versions/b9b68fd74d2a_.py:26]: If you're gonna do this, why not assign random profile pictures?

File: server/app/schema.py [hunk 27]
```diff
@@ -1336,6 +1344,24 @@ def mutate(self, info, **kwargs):
...
+
+    def mutate(self, info, **kwargs):
+
+        picture = kwargs.get("picture")
+        with session_scope() as db_session:
+            curr_user = current_user._get_current_object()
+            if not current_user.is_authenticated:
+                raise GraphQLError(_("You are not logged in."))
+            curr_user.profile_picture = picture
+            return ChangeProfilePicture(ok=True, user=curr_user)
```
Comment (lightyearsaway) [inline server/app/schema.py:1362]: Do we still need ok?

File: server/app/schema.py [hunk 28]
```diff
@@ -1560,6 +1585,14 @@ def resolve_classic(self, info):
         ),
     )
 
+    user_by_name = graphene.Field(User, username=graphene.String(required=True))
+
+    def resolve_user_by_name(self, info, username):
+        return ModelUserAccount.find_by_username(username)
+
+    def resolve_set_count(self):
```
Comment (lightyearsaway) [inline server/app/schema.py:1593]: Hmm, what's this?

File: client/components/common/UserProfile.tsx [hunk 29]
```diff
@@ -0,0 +1,158 @@
...
+        >
+          <img
+            src={getImageUrl(profilePicture)}
+            alt="Avatar"
+            css={{
+              maxWidth: 120,
+              alignItems: 'center',
+              borderRadius: 6,
+              border: '4px solid black',
+              outline: '1px solid #434343',
```
[the remark this answers is not in the document]
Comment (Josemanki) [reply client/components/common/UserProfile.tsx:96]: No specific reason, made it padding and outline has a themed color

File: server/app/schema.py (hunk 28 above)
Comment (Josemanki) [reply server/app/schema.py:1593]: The query to pull the number of sets a user has

File: server/app/schema.py (hunk 27 above)
Comment (Josemanki) [reply server/app/schema.py:1362]: I used it in the modal, preferred a nice ok, haha. But I guess I could check for success with `data?.changeProfilePicture?.user?.profilePicture`

**Comment by lightyearsaway** [comment]
User shouldn’t be nullable, but yeah that’s the general idea. You can assume success if there wasn’t an error.

File: package-lock.json [hunk 30]
```diff
@@ -0,0 +1,6399 @@
+{
```
Comment (lightyearsaway) [inline package-lock.json]: Please delete this file

Comment (Josemanki) [reply package-lock.json]: Deleted

File: client/components/common/UserProfile.tsx [hunk 31]
```diff
@@ -0,0 +1,160 @@
...
+        width: '100%',
+        margin: '0 auto',
+        [mq[1]]: { flexDirection: 'row' },
+      }}
+    >
+      <div
+        css={{
+          display: 'flex',
+          flexDirection: 'row',
+          margin: '0px 20px 20px 0px',
```
Comment (lightyearsaway) [inline client/components/common/UserProfile.tsx:55]: There shouldn't be a right margin on mobile

File: server/app/schema.py [hunk 32]
```diff
@@ -433,6 +433,12 @@ def resolve_custom_sets(self, info, **kwargs):
 
         return query.all()
 
+    custom_set_count = graphene.Int()
```
Comment (lightyearsaway) [inline server/app/schema.py:436]: Could we move this to be under `class NonNullConnection`, like this:

```
    total_count = graphene.Int()

    def resolve_total_count(self, info, **kwargs):
        return len(self.iterable)
```

Then you just add `totalCount` next to `edges` in your query

File: client/components/common/ProfilePictureModal.tsx [hunk 33]
```diff
@@ -0,0 +1,141 @@
...
+          css={{ maxWidth: 220, borderRadius: 4 }}
+        />
+      </div>
+      <div
+        css={{
+          display: 'flex',
+          flexWrap: 'wrap',
+          alignItems: 'center',
+          justifyContent: 'center',
+          gap: '15px',
```
Comment (lightyearsaway) [inline client/components/common/ProfilePictureModal.tsx:96]: Try to use only multiples of 4

File: client/components/common/ProfilePictureModal.tsx [hunk 34]
```diff
@@ -0,0 +1,141 @@
...
+          maxHeight: '30vh',
```
Comment (lightyearsaway) [inline client/components/common/ProfilePictureModal.tsx:97]: Could we find a different way to do this? Maybe limit the modal as a whole? Seems a bit too "magic number" here

File: client/components/common/ProfilePictureModal.tsx [hunk 35]
```diff
@@ -0,0 +1,141 @@
...
+            maxHeight: 'auto',
+          },
+        }}
+      >
+        {PROFILE_PICTURES.map((item, index) => {
+          const activeStyle =
+            active === item
+              ? {
+                  borderRadius: 4,
+                  border: `2px solid ${theme.border?.primarySelected}`,
```
Comment (lightyearsaway) [inline client/components/common/ProfilePictureModal.tsx:109]: prefer box-shadow so we don't affect the placement of anything

File: client/components/common/ProfilePictureModal.tsx [hunk 36]
```diff
@@ -0,0 +1,141 @@
...
+      onCancel();
+    },
+    [onCancel],
+  );
+
+  return (
+    <Modal
+      visible={visible}
+      // Missing translation
+      title={'Change profile picture'}
```
Comment (lightyearsaway) [inline client/components/common/ProfilePictureModal.tsx:73]: Don't forget i18n

File: client/components/common/UserProfile.tsx [hunk 37]
```diff
@@ -0,0 +1,160 @@
...
+          gridArea: '1 / 3 / 1 /1',
+          borderRadius: 6,
```
Comment (lightyearsaway) [inline client/components/common/UserProfile.tsx:57]: I think borderRadius: 4 is fine

File: client/components/common/UserProfile.tsx [hunk 38]
```diff
@@ -0,0 +1,160 @@
...
+          [mq[1]]: {
+            flexDirection: 'column',
+            margin: '10px 0px 0px 0px',
```
Comment (lightyearsaway) [inline client/components/common/UserProfile.tsx:60]: Multiple of 4, if possible?

File: client/components/common/UserProfile.tsx [hunk 39]
```diff
@@ -0,0 +1,160 @@
...
+            padding: '0px 20px 20px 20px',
+            width: '30%',
+            minWidth: '25%',
```
Comment (lightyearsaway) [inline client/components/common/UserProfile.tsx:63]: Why is this necessary? I feel like it could just be 240px?

File: client/components/common/UserProfile.tsx [hunk 40]
```diff
@@ -0,0 +1,160 @@
...
+            },
+          }}
+        >
+          <img
+            src={getImageUrl(profilePicture)}
+            alt="Avatar"
+            css={{
+              maxWidth: 120,
+              alignItems: 'center',
+              borderRadius: 6,
```
Comment (lightyearsaway) [inline client/components/common/UserProfile.tsx:94]: border radius 4

File: client/components/common/UserProfile.tsx [hunk 41]
```diff
@@ -0,0 +1,160 @@
...
+            justifyContent: 'center',
+            marginLeft: 20,
+            [mq[1]]: {
+              marginLeft: 0,
+            },
+          }}
+        >
+          <h1
+            css={{
+              margin: '10px 0px 0px 0px',
```
Comment (lightyearsaway) [inline client/components/common/UserProfile.tsx:138]: multiple of 4

File: client/components/common/UserProfile.tsx [hunk 42]
```diff
@@ -0,0 +1,160 @@
...
+        margin: '0 auto',
+        [mq[1]]: { flexDirection: 'row' },
+      }}
+    >
+      <div
+        css={{
+          display: 'flex',
+          flexDirection: 'row',
+          margin: '0px 20px 20px 0px',
+          gridArea: '1 / 3 / 1 /1',
```
Comment (lightyearsaway) [inline client/components/common/UserProfile.tsx:56]: This is not used, right?

File: client/components/common/ProfilePictureModal.tsx (hunk 34 above)
Comment (Josemanki) [reply client/components/common/ProfilePictureModal.tsx:97]: Hmm, I guess it's doable, but my reasoning behind this wasn't finding a magic number, but for the selector to take a third of the screen. ∼1/3 for spacing on top and bottom, ∼1/3 for current selection, ∼1/3 for gallery. 30-33vh also ensures this ratio is maintained across all mobile screens (all of them have enough height for the component to not look weird too)

File: client/components/common/UserProfile.tsx (hunk 37 above)
Comment (Josemanki) [reply client/components/common/UserProfile.tsx:57]: image
It looks quite pointy, but you're the boss hahaha

File: client/components/common/UserProfile.tsx (hunk 42 above)
Comment (Josemanki) [reply client/components/common/UserProfile.tsx:56]: After having made the change to centered image, no. Removed.

File: client/pages/user/[username].tsx [hunk 43]
```diff
@@ -0,0 +1,123 @@
...
+        <meta property="og:image:type" content="image/png" />
+        <meta
+          property="og:title"
+          content={getTitle(t('USER_PROFILE', { username: username }))}
+        />
+        <meta
+          property="og:description"
+          content={`Discover ${username}'s ${getUserProfileMetaBuildCount(
+            userProfileData?.userByName.customSets.totalCount!,
+          )} build(s) on DofusLab, the open-source set builder for the MMORPG Dofus.`}
```
Comment (lightyearsaway) [inline client/pages/user/[username].tsx:68]: If it's 0 builds, could we fall back to username's profile?

File: client/pages/user/[username].tsx [hunk 44]
```diff
@@ -0,0 +1,123 @@
...
+        />
+        <meta
+          property="twitter:title"
+          content={getTitle(t('USER_PROFILE', { username: username }))}
+        />
+        <meta
+          property="twitter:description"
+          content={`Discover ${username}'s ${getUserProfileMetaBuildCount(
+            userProfileData?.userByName.customSets.totalCount!,
+          )} build(s) on DofusLab, the open-source set builder for the MMORPG Dofus.`}
```
Comment (lightyearsaway) [inline client/pages/user/[username].tsx:78]: Same as above

File: client/pages/user/[username].tsx [hunk 45]
```diff
@@ -0,0 +1,123 @@
...
+            />
+          )}
+        </MobileLayout>
+      </Media>
+      <Media greaterThanOrEqual="xs" css={{ height: '100%' }}>
+        <DesktopLayout showSwitch={false}>
+          {userProfileData?.userByName && (
+            <UserProfile
+              username={username}
+              creationDate={userProfileData?.userByName?.creationDate}
```
Comment (lightyearsaway) [inline client/pages/user/[username].tsx:106]: The `?`s are unnecessary

File: client/pages/user/[username].tsx [hunk 46]
```diff
@@ -0,0 +1,123 @@
...
+import { currentUser } from 'graphql/queries/__generated__/currentUser';
+import ErrorPage from 'pages/_error';
+
+const UserProfilePage: NextPage = () => {
+  const { t } = useTranslation('common');
+
+  const router = useRouter();
+  const username = Array.isArray(router.query.username)
+    ? router.query.username[0]
+    : router.query.username!;
```
Comment (lightyearsaway) [inline client/pages/user/[username].tsx:34]: Not a huge deal, but I'd avoid this by just throwing an error if it doesn't exist.

File: server/app/schema.py [hunk 47]
```diff
@@ -119,6 +122,11 @@ class EdgeBase(graphene.ObjectType, name=f"{node._meta.name}Edge"):
 
         super(NonNullConnection, cls).__init_subclass_with_meta__(node=_node, **kwargs)
 
+    total_count = graphene.Int()
```
Comment (lightyearsaway) [inline server/app/schema.py:125]: required=True, then you can remove the !s after the totalCounts in frontend code
